unit_tests/__init__.py

```diff
@@ -16,7 +16,15 @@
 
 sys.path.append('src')
 sys.path.append('src/lib')
+sys.path.append('src/actions')
 
 # Mock out charmhelpers so that we can test without it.
 import charms_openstack.test_mocks  # noqa
 charms_openstack.test_mocks.mock_charmhelpers()
+
+import mock
+import charms
+keystoneauth1 = mock.MagicMock()
+sys.modules['keystoneauth1'] = keystoneauth1
+charms.leadership = mock.MagicMock()
+sys.modules['charms.leadership'] = charms.leadership
```

Comment:
When I've directly inserted mock objects into sys.modules, I've tended to regret it. They don't get reset between tests, so you can wind up polluting one test run with data from another.

I'd recommend instead using the mock.patch start and stop (and maybe reset) commands in your setUp and tearDown methods. (You can also use setUpClass and tearDownClass if you want to wrap the run of every test in a class ... though again, you can run into data pollution issues).

Reply:
Pete, this is a method we have used in many charms. Specifically, it tends to be modules which are not getting tested directly but which may have side effects.

Reply:
Sorry to chime in here, but:

Mocking in `unit_tests/__init__.py` was *never* meant for actual unit tests, but simply to keep Python happy during import time of the modules *under* tests; i.e. as David says, "... which may have side effects".  Explicitly, with reactive charms, it's to work around the 'self healing' that charm-helpers employs using `import module -> fail -> apt install python-module -> import module`.

Mocks in the `unit_tests/__init__.py` should *never* be assigned return values, and then checked in tests.  That way leads to self harm ... re-mocking is totally appropriate in the unit-test, and it makes it explicit what the unit test and the function-under-test needs.

Reply:
As a fix for self healing, I understand. It's a dangerous solution, but it is a solution. I'd prefer a comment to that effect wherever we do the import, so people know not to use those values in tests.